### PR TITLE
feat(iOS): Add support for UINavigationBackButtonDisplayMode

### PR DESCRIPTION
**Pull request by maciekstosio**
## Description

~This PR improves upon #2105. #2105 allowed to use iOS 14 default back button behavior when label is not provided. This PR allows to modify the behavior by allowing to provide UINavigationButtonBackButtonDisplayMode and enables it for custom text (without style modifications). The main problem is that we used to provide backButtonItem in most of the cases which [disables](https://developer.apple.com/documentation/uikit/uinavigationitem/3656350-backbuttondisplaymode) backButtonDisplayMode.~

This PR adds possibility to customize default behavior of back button using `backButtonDisplayMode` ([UINavigationBackButtonDisplayMode](https://developer.apple.com/documentation/uikit/uinavigationitem/backbuttondisplaymode)) for iOS.

:warning: **This modifies only default back button**, when any customization is added (including headerBackTitle) in native part we create custom `RNSUIBarButtonItem` and set it as `backButtonItem`, which [disables](https://developer.apple.com/documentation/uikit/uinavigationitem/3656350-backbuttondisplaymode) `backButtonDisplayMode` behavior.

I tried to make it work together with custom label (`headerBackTitle`) using `prevItem.backButtonTitle`, but due to iOS limitations it is not viable option. It influences also back button menu - changes the label of previous screen - which is not the behavior we want. 

To sum up, `backButtonDisplayMode` work when none of:
- `headerBackTitleStyle.fontFamily`
- `headerBackTitleStyle.fontSize`
- `headerBackTitle`
- `disableBackButtonMenu`

are set. 

## Screenshots / GIFs

|Paper|Fabric|
|-|-|
|<video src="https://github.com/software-mansion/react-native-screens/assets/11800297/c6aa7697-4331-4cb4-a81d-7f77f128513d" />|<video src="https://github.com/software-mansion/react-native-screens/assets/11800297/fa0edd92-1aa2-45e5-a466-516c0ec120d2" />|

<details>
<summary>Example component used in tests:</summary>

```jsx
import * as React from 'react';
import { Button, View, Text, StyleSheet } from 'react-native';
import { NavigationContainer, ParamListBase } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';

const Stack = createNativeStackNavigator();

type NavProp = {
  navigation: NativeStackNavigationProp<ParamListBase>;
};

export default function App() {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          name="screenA"
          component={ScreenA}
          options={{ headerTitle: 'A: Home' }}
        />
        <Stack.Screen
          name="screenB"
          component={ScreenB}
          options={{
            headerTitle: 'B: default',
            backButtonDisplayMode: 'default',
          }}
        />
        <Stack.Screen
          name="screenC"
          component={ScreenC}
          options={{
            headerTitle: 'C: generic',
            backButtonDisplayMode: 'generic',
          }}
        />
        <Stack.Screen
          name="screenD"
          component={ScreenD}
          options={{
            headerTitle: 'D: minimal',
            backButtonDisplayMode: 'minimal',
          }}
        />
        <Stack.Screen
          name="screenE"
          component={ScreenE}
          options={{
            headerTitle: 'E: custom',
            headerBackTitle: 'Back Title',
            backButtonDisplayMode: 'minimal',
          }}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
}

const ScreenA = ({ navigation }: NavProp) => (
  <View style={styles.container}>
    <Text>Screen A</Text>
    <Button
      onPress={() => navigation.navigate('screenB')}
      title="Go to screen B"
    />
  </View>
);

const ScreenB = ({ navigation }: NavProp) => (
  <View style={styles.container}>
    <Text>Screen B</Text>
    <Text>backButtonDisplayMode: default</Text>
    <Button
      onPress={() => navigation.navigate('screenC')}
      title="Go to screen C"
    />
  </View>
);

const ScreenC = ({ navigation }: NavProp) => (
  <View style={{ flex: 1, paddingTop: 50 }}>
    <Text>Screen C</Text>
    <Text>backButtonDisplayMode: generic</Text>
    <Button
      onPress={() => navigation.navigate('screenD')}
      title="Go to screen D"
    />
  </View>
);

const ScreenD = ({ navigation }: NavProp) => (
  <View style={styles.container}>
    <Text>Screen D</Text>
    <Text>backButtonDisplayMode: minimal</Text>
    <Button
      onPress={() => navigation.navigate('screenE')}
      title="Go to screen E"
    />
  </View>
);

const ScreenE = (_props: NavProp) => (
  <View style={styles.container}>
    <Text>Screen E</Text>
    <Text>backButtonDisplayMode omitted because of the headerBackTitle</Text>
  </View>
);

const styles = StyleSheet.create({
  container: { flex: 1, alignItems: 'center', justifyContent: 'space-around' },
});
```

</details>

## Checklist

- [x] Included code example that can be used to test this change
- [x] Updated TS types
- [x] Updated documentation: 
  - [x] https://github.com/software-mansion/react-native-screens/blob/main/guides/GUIDE_FOR_LIBRARY_AUTHORS.md
  - [x] https://github.com/software-mansion/react-native-screens/blob/main/native-stack/README.md
  - [x] https://github.com/software-mansion/react-native-screens/blob/main/src/types.tsx
  - [x] https://github.com/software-mansion/react-native-screens/blob/main/src/native-stack/types.tsx
- [x] Ensured that CI passes

Tested #1864: Paper ✅ Fabric ✅
Tested #1646: Paper ❌ Fabric ❌ - but it does not work on main too, could now be achieved using `backButtonDisplayMode: ‘minimal’`

### CODE REVIEW

File: Example/ios/Podfile.lock [hunk 1]
```diff
@@ -435,9 +435,10 @@ PODS:
     - RCT-Folly (= 2021.07.22.00)
     - React-Core
     - ReactCommon/turbomodule/core
-  - RNScreens (3.31.0-rc.1):
+  - RNScreens (3.31.1):
```
Comment (kkafar) [inline Example/ios/Podfile.lock:438]: This is fine, however it would be nice to prepare separate PR with update of podfiles in all examples. Usually we try to not include changes in Podfile.lock & yarn.lock in regular PRs, but this is only a thumb rule.

File: android/src/main/java/com/swmansion/rnscreens/ScreenStackHeaderConfigViewManager.kt [hunk 2]
```diff
@@ -201,4 +201,8 @@ class ScreenStackHeaderConfigViewManager : ViewGroupManager<ScreenStackHeaderCon
     override fun setDisableBackButtonMenu(view: ScreenStackHeaderConfig?, value: Boolean) {
         logNotAvailable("disableBackButtonMenu")
     }
+
+    override fun setBackButtonDisplayMode(view: ScreenStackHeaderConfig?, value: String?) {
+        logNotAvailable("backButtonDisplayMode")
+    }
```
Comment (kkafar) [inline android/src/main/java/com/swmansion/rnscreens/ScreenStackHeaderConfigViewManager.kt:207]: Does it work fine? Because I do not see `RNSScreenStackHeaderConfigManagerInterface<ScreenStackHeaderConfig>` being updated.

What we need to do here is run this project on Android Fabric (or just trigger the codegen manually from Android Studio) & copy generated `RNSScreenStackHeaderConfigManagerInterface<ScreenStackHeaderConfig>` interface from build folders `android/build/generate/source/codegen/java/com/facebook/react/viewmanagers/...` to `android/src/paper/java/com/facebook/react/viewmanagers/` so that the interfaces are up to date.

TODO: We need to automate this somehow.

File: ios/RNSConvert.h [hunk 3]
```diff
@@ -6,6 +6,13 @@ namespace react = facebook::react;
 
 @interface RNSConvert : NSObject
 
++ (NSNumber *)getFontSizePropValue:(int)value;
+
++ (UISemanticContentAttribute)getDirectionPropValue:(react::RNSScreenStackHeaderConfigDirection)direction;
+
++ (UINavigationItemBackButtonDisplayMode)getBackButtonDisplayModePropValue:
+    (react::RNSScreenStackHeaderConfigBackButtonDisplayMode)backButtonDisplayMode;
```
Comment (kkafar) [inline ios/RNSConvert.h:14]: Would suggest sticking to convention here:
```suggestion
+ (UINavigationItemBackButtonDisplayMode)UINavigationItemBackButtonDisplayModeFromCppEquivalent:
    (react::RNSScreenStackHeaderConfigBackButtonDisplayMode)backButtonDisplayMode;
```

File: ios/RNSConvert.h [hunk 4]
```diff
@@ -6,6 +6,13 @@ namespace react = facebook::react;
 
 @interface RNSConvert : NSObject
 
++ (NSNumber *)getFontSizePropValue:(int)value;
```
Comment (kkafar) [inline ios/RNSConvert.h:9]: Let's leave this particular method in the component implementation file as is not straight up conversion but logic specific to the particular component. 

Alternatively we could name it `NSNumberFromPositiveIntOrNil` or something to make clear what it does.

File: ios/RNSConvert.h [hunk 5]
```diff
@@ -6,6 +6,13 @@ namespace react = facebook::react;
...
+
++ (UISemanticContentAttribute)getDirectionPropValue:(react::RNSScreenStackHeaderConfigDirection)direction;
```
Comment (kkafar) [inline ios/RNSConvert.h:11]: The same recommendation here: let's try to stick to naming convention.

File: ios/RNSConvert.h (hunk 3 above)
Comment (kkafar) [reply ios/RNSConvert.h:14]: And it would be great to move these in separate PR. But this time it is fine I guess.

File: ios/RNSScreenStackHeaderConfig.mm [hunk 6]
```diff
@@ -511,7 +512,13 @@ + (void)updateViewController:(UIViewController *)vc
                                                                              action:nil];
   [backBarButtonItem setMenuHidden:config.disableBackButtonMenu];
 
-  auto isBackButtonCustomized = !isBackTitleBlank || config.disableBackButtonMenu;
+  auto overrideBackButtonItem =
+      config.disableBackButtonMenu; // when config.disableBackButtonMenu is true we need to set backButtomItem
+  prevItem.backButtonTitle = resolvedBackTitle;
```
Comment (kkafar) [inline ios/RNSScreenStackHeaderConfig.mm:517]: Hmm, now we do always set `backButtonTitle` of `prevItem` 🤔 We need to test it thoroughly for regressions. We need to prepare list of tests examples to test before we can process further.

File: src/native-stack/types.tsx [hunk 7]
```diff
@@ -112,6 +112,15 @@ export type NativeStackNavigationOptions = {
...
   disableBackButtonMenu?: boolean;
+  /**
+   * How the back button behaves by default (when not customized). Available on iOS>=14.
+   * The following values are currently supported (they correspond to https://developer.apple.com/documentation/uikit/uinavigationitembackbuttondisplaymode?language=objc):
+   * - "default" – shows given back button title/previous controller title, system default or just icon based on available space
+   * - "generic" – shows given system default or just icon based on available space
+   * - "minimal" – shows just an icon
+   * @platform ios
+   */
+  backButtonDisplayMode?: ScreenStackHeaderConfigProps['backButtonDisplayMode'];
```
Comment (kkafar) [inline src/native-stack/types.tsx:123]: We need to also think through how this property interacts with `backTitleVisible` prop. 
What happens if we set `backTitleVisible: false` and set this to default or other options? And the other way around.

Comment (kkafar) [reply src/native-stack/types.tsx:123]: At least we should document these interactions between any related props here.

**Review by kkafar — COMMENTED** [review]
Great job 🎉 I've left few comments, please answer them.

File: Example/ios/Podfile.lock (hunk 1 above)
Comment (maciekstosio) [reply Example/ios/Podfile.lock:438]: I'll prepare separate PR with bump

File: src/native-stack/types.tsx (hunk 7 above)
Comment (maciekstosio) [reply src/native-stack/types.tsx:123]: I don't think that the logic changes significantly in this case, pls look at:
image
This is how it worked before, we set prevItem.backButtonTitle for back button menu when backTitleVisible was false.

File: Example/ios/Podfile.lock (hunk 1 above)
Comment (maciekstosio) [reply Example/ios/Podfile.lock:438]: And revert this change here.

File: src/native-stack/types.tsx (hunk 7 above)
Comment (maciekstosio) [reply src/native-stack/types.tsx:123]: Ok, I actually revert bunch of this changes. We can't use backButtonTitle to set custom label - more details in PR summary.

File: ios/RNSScreenStackHeaderConfig.mm (hunk 6 above)
Comment (maciekstosio) [reply ios/RNSScreenStackHeaderConfig.mm]: I reverted that change, seems like it could have negative impact on back button menu.

File: Example/ios/Podfile.lock (hunk 1 above)
Comment (maciekstosio) [reply Example/ios/Podfile.lock]: Bump in #2133

**Review by tboba — APPROVED** [review]
LGTM! Nice catch with that `getDimensionPropValue` 🎉

File: ios/RNSScreenStackHeaderConfig.mm [hunk 8]
```diff
@@ -513,6 +514,10 @@ + (void)updateViewController:(UIViewController *)vc
 
   auto isBackButtonCustomized = !isBackTitleBlank || config.disableBackButtonMenu;
 
+  if (@available(iOS 14.0, *)) {
+    prevItem.backButtonDisplayMode = config.backButtonDisplayMode;
+  }
```
Comment (kkafar) [inline ios/RNSScreenStackHeaderConfig.mm:519]: I think we need ifdef checks for iOS version here.
Look for ifdefs for iOS 15 examples in code.

File: ios/RNSScreenStackHeaderConfig.mm [hunk 9]
```diff
@@ -945,7 +942,9 @@ - (UIView *)view
 RCT_EXPORT_VIEW_PROPERTY(hideShadow, BOOL)
 RCT_EXPORT_VIEW_PROPERTY(backButtonInCustomView, BOOL)
 RCT_EXPORT_VIEW_PROPERTY(disableBackButtonMenu, BOOL)
-// `hidden` is an UIView property, we need to use different name internally
+RCT_EXPORT_VIEW_PROPERTY(
+    backButtonDisplayMode,
+    UINavigationItemBackButtonDisplayMode) // `hidden` is an UIView property, we need to use different name internally
```
Comment (kkafar) [inline ios/RNSScreenStackHeaderConfig.mm:947]: Let's fixup this comment here, it should be on `hidden` prop line.

File: native-stack/README.md [hunk 10]
```diff
@@ -79,6 +79,10 @@ String that applies `rtl` or `ltr` form to the stack. On Android, you have to ad
 
 Boolean indicating whether to show the menu on longPress of iOS >= 14 back button.
 
+#### `backButtonDisplayMode` (iOS only)
+
+Enum value indicating display mode of **default** back button. It works on iOS >= 14, and is used only when none of:  `backTitleFontFamily`, `backTitleFontSize`, `disableBackButtonMenu` or `backTitle` is set. Otherwise, when the button is customized, under the hood we use iOS native `backButtonItem` which overrides `backButtonDisplayMode`. Read more [#2123](https://github.com/software-mansion/react-native-screens/pull/2123).
```
Comment (kkafar) [inline native-stack/README.md:84]: Here we also lack information on what values are accepted.

File: src/native-stack/types.tsx [hunk 11]
```diff
@@ -112,6 +112,15 @@ export type NativeStackNavigationOptions = {
    * @platform ios
    */
   disableBackButtonMenu?: boolean;
+  /**
+   * How the back button behaves by default (when not customized). Available on iOS>=14, and is used only when none of: `backTitleFontFamily`, `backTitleFontSize`, `disableBackButtonMenu` or `backTitle` is set.
+   * The following values are currently supported (they correspond to https://developer.apple.com/documentation/uikit/uinavigationitembackbuttondisplaymode?language=objc):
+   * - "default" – show given back button previous controller title, system generic or just icon based on available space
+   * - "generic" – show given system generic or just icon based on available space
+   * - "minimal" – show just an icon
+   * @platform ios
```
[the remark this answers is not in the document]
Comment (tboba) [reply src/native-stack/types.tsx:121]: I don't know if suggesting `iOS` is good idea right now, since we're doing `@platform android` / `@platform ios` everywhere. Let's maybe do that in separate PR.

[the remark this answers is not in the document]
Comment (maciekstosio) [reply src/native-stack/types.tsx:121]: I agree with @tboba almost always we use `@platform ios` there is just one case when we write `iOS`

[the remark this answers is not in the document]
Comment (kkafar) [reply src/native-stack/types.tsx:121]: That's a subject to change then, but it might be done in separate PR, I do not mind.

**Review by kkafar — APPROVED** [review]
I think we're good to go here 🎉
